leader-election-impl/src/test/java/com/palantir/leader/proxy/AwaitingLeadershipProxyTest.java

```diff
@@ -196,6 +197,23 @@ public void shouldBlockOnGainingLeadershipIfNotCurrentlyLeading() throws Excepti
         verify(leaderElectionService).blockOnBecomingLeader();
     }
 
+    @Test
+    public void shouldRetryBecomingLeader() throws Exception {
+        when(leaderElectionService.blockOnBecomingLeader())
+                .thenThrow(new RuntimeException())
+                .thenReturn(leadershipToken);
+
+        Runnable proxy = AwaitingLeadershipProxy.newProxyInstance(
+                Runnable.class,
+                delegateSupplier,
+                leaderElectionService);
+
+        Thread.sleep(1000); //wait for retrying on gaining leadership
```

Comment:
hmm, the `Thread.sleep` here and also in the impl does make your test hard to write and not flake.

If we have to do the `Thread.sleep` let's wait for 2 seconds and then instead of verifying exactly twice, just say `leaderElectionService.blockOnBecomingLeader()` happens at least twice, which is the behaviour you want.

Alternatively, the executor approach would get around this + package private constructor we use just for tests 👀. Not blocking on this, just the former.